### ISSUE TITLE
Awful quantile estimation on small sample size

### ISSUE THREAD

**Issue by caio** [caio/go-tdigest#12]
Following the discussion on PR #11 I've isolated a tiny pathological sample:

```go
t := New(10)
data := []float64{0, 279, 2, 281}
for _, f := range data {
    t.Add(f, 1)
}
fmt.Println(t.Quantile(0.25))
```
Which yields `-67.7500`. In hindsight, this outrageously wrong number should have raised a major red flag in my head already, but when comparing to Java this gets even more obviously wrong:

```java
public class Ughhh {
    public static void main(String[] args) {
        TDigest t = new AVLTreeDigest(10);
        Arrays.asList(0, 279, 2, 281).forEach(t::add);
        System.out.println(t.quantile(0.25));
    }
}
```

Prints out `1.0` 😅 (Note: before the min/max patch on https://github.com/tdunning/t-digest/commit/89bb394dca674cad5b6058ebd300ff0e64748c0f it yields `1.5`).

**Comment by caio** [comment]
So here's my current understanding of this issue:

This implementation estimates the value under a quantile following `Algorithm 4` described [in the paper][paper] whereas the reference implementation [uses a different method][impl].

Whilst using the reference implementation does behave a lot better for this pathological example, it also [suffers from a serious bug][bug] (Notice that the test is [disabled in the reference implementation][disabled]).

Interestingly the current version does *not* suffer from the `singletonInACrowd` bug. So it would be trading one bug for another. It's hard for me to say which is worse for I don't quite understand the impact of this current issue on similar-but-real shapes.

Apparently [there's a new approach being developed][newavl] which I'll take a look at as soon as possible. I also skimmed through the merging digest implementation and it seems like it's not much work to port it.

[paper]: https://github.com/tdunning/t-digest/blob/master/docs/t-digest-paper/histo.pdf
[impl]: https://github.com/tdunning/t-digest/blob/641646ffa47420c28b17c4ae71e13ffc507deb25/core/src/main/java/com/tdunning/math/stats/AVLTreeDigest.java#L256
[bug]: https://github.com/caio/go-tdigest/commit/743df6582d8bf213ac0ea4d90b6443fc136c2ae9
[disabled]: https://github.com/tdunning/t-digest/blob/641646ffa47420c28b17c4ae71e13ffc507deb25/core/src/test/java/com/tdunning/math/stats/AVLTreeDigestTest.java#L46
[newavl]: https://github.com/tdunning/t-digest/blob/issue-84/src/main/java/com/tdunning/math/stats/AVLTreeDigest.java#L237

**Comment by caio** [comment]
Alright, branch `avl-wip` now contains code that is able to close this issue.

Gotchas (explanation in the commit message):

* https://github.com/caio/go-tdigest/commit/8a6b3a592c18f732dbffeb383d6f1b3f2568bd33
* https://github.com/caio/go-tdigest/commit/baa019061e8432a365c9c77e025459c44b79b46e

Another notable commit (because it makes test execution significantly slower): https://github.com/caio/go-tdigest/commit/79b62e94c068048bbdb5350c7628ad86aeb01f1b

I'm pretty happy with the results, even considering the above gotchas so my plan is cleaning the branch up (a lot of code has gone away, I didn't pay attention to abstractions nor "niceties") when I can and merging it as a real fix for this. If anyone is watching, now it's the time to shout 😝

**Comment by caio** [comment]
The fix has been released on v1.1.3: https://github.com/caio/go-tdigest/releases/tag/v1.1.3
